cmake/kokkos_tribits.cmake

```diff
@@ -241,6 +241,37 @@ function(KOKKOS_SET_LIBRARY_PROPERTIES LIBRARY_NAME)
     target_link_options(${LIBRARY_NAME} PUBLIC ${KOKKOS_LINK_OPTIONS})
   endif()
 
+  if(CMAKE_VERSION VERSION_GREATER_EQUAL 3.18)
+    if(KOKKOS_LINK_OPTIONS)
+      include(CheckLinkerFlag)
+      #exclude case of compiler_launcher forwarding to nvcc_wrapper as the used CXX compiler is shadowed in this case (compiler_launcher changes the compiler).
+      #The CXX compiler CMake will invoke for the check is not able to consume the cuda flags.
+      if(NOT (KOKKOS_ENABLE_CUDA) OR ("${CMAKE_CXX_COMPILER}" MATCHES "nvcc_wrapper") OR (${KOKKOS_CXX_COMPILER_ID}
+                                                                                          STREQUAL Clang)
+      )
+        #check_linker_flag requires a whitespace separated list
+        string(REPLACE ";" " " WHITESPACE_FLAGS "${KOKKOS_LINKER_OPTIONS}")
+        # sycl adds "-device ..." options that need quotes (which CMake removes). We need to add them here again.
+        string(REGEX REPLACE "(-device [A-Za-z0-9_\\\\.]*)" "\"\\1\"" QUOTED_FLAGS "${WHITESPACE_FLAGS}")
+        # temporarily set language flags to nothing ... the linker often can not handle these which leads to false errors
+        set(FLAGS_CACHE CMAKE_${KOKKOS_COMPILE_LANGUAGE_FLAGS})
+        set(CMAKE_${KOKKOS_COMPILE_LANGUAGE}_FLAGS "")
+        #disable caching
+        unset(KOKKOS_LINK_OPTIONS_CHECK CACHE)
+        check_linker_flag(${KOKKOS_COMPILE_LANGUAGE} "${QUOTED_FLAGS}" KOKKOS_LINK_OPTIONS_CHECK)
+        #restore cache
+        set(CMAKE_${KOKKOS_COMPILE_LANGUAGE}_FLAGS "${FLAGS_CACHE}")
```

Comment:
Aren't `CMAKE_${KOKKOS_COMPILE_LANGUAGE}_FLAGS` `CACHE` variables and setting them requires the `CACHE` keyword?

Reply:
good catch

Reply:
definitely some parenthesis and quotes :-D ... love cmake

Reply:
It's not quite clear to me whether `check_linker_flag` uses the non-CACHE or CACHE variable `CMAKE_CXX_FLAGS`. Since it  seems to make a difference setting the non-CACHE variable here, it's probably just looking for `CMAKE_CXX_FLAGS` picking up the non-CACHE one if it exists. Since we are indeed in a function here, there doesn't seem to be any need to reset the non-CACHE variable (and thus we don't need `FLAGS_CACHE`).

Reply:
`check_linker_flag` uses the non-Cache var (if I disable setting it to "" it fails since stuff like -Werror makes it fail for some configs).
In `KOKKOS_SET_LIBRARY_PROPERTIES` we do `check_compiler_flag` after the linker flag check. Since we want this check to have the CXX flags, we need to reset it after `check_linker_flag`. But we could move the linker check to the end and only disable the cache there.

Reply:
OK. I missed that `check_compiler_flag` is also part of the same function. Also, we are setting the non-CACHE variable already for `Kokkos_ENABLE_COMPILER_WARNINGS` (since we don't want to change the users flags but still not export the warning flags) so it makes sense to cache and restore.

Reply:
Do you have any opinions on putting `check_linker_flag` to the end? Honestly the order does not have any real impact on the functionality but it would get rid of the cache restore

Reply:
No, the order doesn't matter but we should have the cache restore anyway (because I don't want to rely on any assumptions when `CMAKE_CXX_FLAGS` is used).

Reply:
ok, then I will just leave it where it is. I think it makes sense to see that the flags are checked directly after they are added